feature/141, OCPP 1.6 Security White Paper Ed 2

villekr opened the pull request:
Improved Security for OCPP 1.6-J specification (https://www.openchargealliance.org/news/enhanced-security-for-ocpp-16/) adds few messages backported from 2.0 to 1.6. Some messages have been renamed to prevent conflict with original 1.6 messages. Also some schemas have modified content. Open Charge Alliance hasn't released official backported schema files.

Changes to mobilityhouse/ocpp:
- Copied selected 2.0 schema files to 1.6. Files are stored ocpp/v16/schemas/backport to clearly separate them from original 1.6 files.
- Backported schema files have been renamed according to improved security specification, and their content has been checked and modified.
- Added support for loading backported schema files.
- Added test case to verify that using backported action type works.

This PR is a fix for https://github.com/mobilityhouse/ocpp/issues/141.

OrangeTux commented on ocpp/messages.py:
```diff
@@ -182,6 +182,53 @@ def get_validator(
+    except FileNotFoundError:
+        return _get_backport_validator(
+            message_type_id,
+            action,
+            ocpp_version,
+            parse_float
+        )
+
+
+def _get_backport_validator(
+        message_type_id: int,
+        action: str,
+        ocpp_version: str,
+        parse_float: Callable = float
+) -> Draft4Validator:
+    """
+    Reads backported schema from disk.
+
+    OCPP 1.6 Improved Security specification has selected Requests and
+    Responses backported from 2.0 to 1.6.
+    """
+    if ocpp_version not in ["1.6"]:
+        raise ValueError
+
+    schemas_dir = 'v' + ocpp_version.replace('.', '')
+
+    schema_name = action
+    if message_type_id == MessageType.CallResult:
+        schema_name += 'Response'
+    elif message_type_id == MessageType.Call:
+        schema_name += 'Request'
+
+    cache_key = schema_name + '_' + ocpp_version
+    if cache_key in _validators:
+        return _validators[cache_key]
+
+    dir,  _ = os.path.split(os.path.realpath(__file__))
+    relative_path = f'{schemas_dir}/schemas/backport/{schema_name}.json'
+    path = os.path.join(dir, relative_path)
+
```
This changes introduce duplicate code to open and parse the schema's and to create a validator. This can be avoided by putting the schemas inside the same folder as all other OCPP 1.6 schema's. Could you remove this duplication?

OrangeTux commented on ocpp/v16/call.py:
```diff
@@ -229,3 +234,72 @@ class DataTransferPayload:
+
+@dataclass
+class CertificateSignedPayload:
+    cert: List
```
The schemas for this request don't match the specification:

![Screenshot_2021-04-26_11-16-26](https://user-images.githubusercontent.com/1565144/116059259-f0bee880-a680-11eb-97d0-16714d2593e0.png)

My suggestion is to follow the written specification. Please use `certificate_chain` in both here and the schema's.

OrangeTux replied:
Make sure to update the type from `List` to `str` as well.

OrangeTux commented on ocpp/v16/call.py:
```diff
@@ -229,3 +234,72 @@ class DataTransferPayload:
+
+@dataclass
+class ExtendedTriggerMessagePayload:
+    requested_message: MessageTriggerType
```
the field `connetor_id` is missing. Please add it.

![Screenshot_2021-04-26_11-19-21](https://user-images.githubusercontent.com/1565144/116059572-44313680-a681-11eb-871f-785dd6c5bca3.png)

OrangeTux commented on ocpp/v16/call.py:
```diff
@@ -229,3 +234,72 @@ class DataTransferPayload:
+
+@dataclass
+class GetInstalledCertificateIdsPayload:
+    type_of_certificate: CertificateUseType
```
Please use the field name `certificate_type`.

![Uploading Screenshot_2021-04-26_11-20-00.png…]()

OrangeTux commented on ocpp/v16/call.py:
```diff
@@ -229,3 +234,72 @@ class DataTransferPayload:
+@dataclass
+class SecurityEventNotificationPayload:
+    type: str
+    timestamp: str
+
```
Please add the missing field `tech_info`.
![Uploading Screenshot_2021-04-26_11-22-18.png…]()

OrangeTux reviewed (changes requested):
Thanks for your PR. I've some suggestion for improvements. 

I don't like the suffix `Type` to all enums. although it's according to the specification. So I'm not sure yet if we should keep it or remove it. I'll come back to you regarding this topic.

villekr commented:
I've made the following changes:
- Replaced backported schema files with official ones (https://www.openchargealliance.org/uploads/files/OCPP-1.6-security-whitepaper-edition-2.zip) instead of the ones copied and modified form v20/schemas folder.
- Stored backported schema files in v16/schemas. Removed related code changes that were added to messages.py to read schemas from v16/schemas/backport.
- Renamed enum types without Type-suffix so naming now follows the pattern (type in spec) CertificateSignedStatusEnumType -> (type in code) CertificateSignedStatus. The following ones are exceptions due conflict in original types: SignedFirmwareStatus, ExtendedMessageTrigger, ExtendedTriggerMessageStatus.

tropxy commented on ocpp/messages.py:
```diff
@@ -182,6 +182,53 @@ def get_validator(
+    Responses backported from 2.0 to 1.6.
+    """
+    if ocpp_version not in ["1.6"]:
+        raise ValueError
```
Can you add context to the ValueError. Otherwise it will be harder to understand what was the error exactly

villekr replied:
This was my mistake. I thought I had removed the _get_backport_validator but apparently did some mistake and it was still there.

tropxy commented on ocpp/messages.py:
```diff
@@ -182,6 +182,53 @@ def get_validator(
+    if message_type_id == MessageType.CallResult:
+        schema_name += 'Response'
+    elif message_type_id == MessageType.Call:
+        schema_name += 'Request'
```
The new schemas for 1.6 still have the same format, so we dont need to add the Request to the name. Which brings to my second point, that I dont thing we need an extra backport code to extract the new schemas. we can simply reuse the old code for that purpose

villekr replied:
Same as above. I'll remove this logic.

tropxy commented on ocpp/v16/enums.py:
```diff
@@ -31,6 +31,18 @@ class Action(str, Enum):
+    SecurityEventNotification = "SecurityEventNotification"
+    SignCertificate = "SignCertificate"
+    SignedFirmwareStatusNotification = "SignedFirmwareStatusNotification"
+    SignedUpdateFirmware = "UpdateFirmware"
```
I guess it was supposed to be SignedUpdateFirmware

tropxy replied:
Also in here: https://github.com/mobilityhouse/ocpp/pull/206/files#diff-8d8918e730a141086aa194d66dabdf5b389795650190b713f7a2e0b29d8ce04dL287

the following fields are missing:
* DownloadScheduled
* DownloadPaused
* InstallRebooting
* InstallScheduled
* InstallVerificationFailed
* InvalidSignature
* SignatureVerified

we also have repeated this fields, so please, if you dont mind could you remove them? @OrangeTux do we know why we have these fields repeated?
https://github.com/mobilityhouse/ocpp/blob/8b055578660c96740df99d4e2614e8157b51929b/ocpp/v16/enums.py#L312

villekr replied:
This comment is a little unclear for me. Enum values for FirmwareStatusNotificationPayload must be different compared to SignedFirmwareStatus. 

@dataclass
class FirmwareStatusNotificationPayload:
    status: FirmwareStatus

class FirmwareStatus(str, Enum):
    """
    Status of a firmware download as reported in FirmwareStatusNotification.req
    """

    downloaded = "Downloaded"
    download_failed = "DownloadFailed"
    downloading = "Downloading"
    idle = "Idle"
    installation_failed = "InstallationFailed"
    installing = "Installing"
    installed = "Installed"

@dataclass
class SignedFirmwareStatusNotificationPayload:
    status: SignedFirmwareStatus
    request_id: int

----

class SignedFirmwareStatus(str, Enum):
    """
    FirmwareStatusEnumType is used by: SignedFirmwareStatusNotification.req
    """

    certificate_verified = "CertificateVerified"
    downloaded = "Downloaded"
    download_failed = "DownloadFailed"
    downloading = "Downloading"
    download_scheduled = "DownloadScheduled"
    download_paused = "DownloadPaused"
    idle = "Idle"
    installation_failed = "InstallationFailed"
    installing = "Installing"
    installed = "Installed"
    install_rebooting = "InstallRebooting"
    install_scheduled = "InstallScheduled"
    install_verification_failed = "InstallVerificationFailed"
    invalid_signature = "InvalidSignature"
    invalid_certificate = "InvalidCertificate"
    revoked_certificate = "RevokedCertificate"
    signature_verified = "SignatureVerified"

tropxy replied:
I didnt notice that class. However, I must say you are basing your code in a outdated version of the whitepaper. I uploaded here both the spec and the schemas
[OCPP 1.6 security whitepaper edition 2.zip](https://github.com/mobilityhouse/ocpp/files/6435252/OCPP.1.6.security.whitepaper.edition.2.zip)

tropxy replied:
Nevertheless, again the Spec doesnt distinguish between SignedFirmwareStatusEnum and FirmwareStatusEnum, because they share the same core, i.e., adding more options to FirmwareStatusEnum is still backwards compatible with a plain OCPP 1.6J version.

villekr replied:
You're definitely right. Removed 'InvalidCertificate', 'RevokedCertificate', 'CertificateVerified' enums and added new options to FirmwareStatus and removed SignedFirmwareStatus.

tropxy commented on ocpp/v16/enums.py:
```diff
@@ -619,3 +631,156 @@ class ValueFormat(str, Enum):
+    status_notification = "StatusNotification"
+
+
+class ExtendedTriggerMessageStatus(str, Enum):
+    """
+    TriggerMessageStatusEnumType is used by: ExtendedTriggerMessage.conf
+    """
+
+    accepted = "Accepted"
+    rejected = "Rejected"
+    not_implemented = "NotImplemented"
+
```
This already exists as TriggerMessageStatus, so I dont think we need to repeat it with a different name. Also int the whitepaper is defined as "TriggerMessageStatus" and I think we should stick with the whitepaper terms

villekr replied:
That's a good point as values are the same there's no point duplicating this type.

tropxy commented on ocpp/v16/enums.py:
```diff
@@ -619,3 +631,156 @@ class ValueFormat(str, Enum):
+    accepted_canceled = "AcceptedCanceled"
+
+
+class ExtendedMessageTrigger(str, Enum):
+    """
+    MessageTriggerEnumType is used by: ExtendedTriggerMessage.req
+    """
+
+    boot_notification = "BootNotification"
+    log_status_notification = "LogStatusNotification"
+    firmware_status_notification = "FirmwareStatusNotification"
+    heartbeat = "Heartbeat"
+    meter_values = "MeterValues"
+    sign_charge_point_certificate = "SignChargePointCertificate"
+    status_notification = "StatusNotification"
```
I dont think we need to create a new Status for the Extended Trigger. We can just add the extra fields that are added by the whitepaper and add a comment mentioning it. This will then be in accordance with the Whitepaper which also doesnt change the name of the enum "TriggerMessageStatusEnumType"

villekr replied:
I'll make that change also.

tropxy commented on ocpp/v16/enums.py:
```diff
@@ -619,3 +631,156 @@ class ValueFormat(str, Enum):
+    """
+
+    accepted = "Accepted"
+    rejected = "Rejected"
+    accepted_canceled = "AcceptedCanceled"
+
```
These are missing:
* InvalidCertificate
* RevokedCertificate

villekr replied:
Well spotted. These seem to be missing from edition 2 FINAL, 2017-09-28 specification but do exists in SignedUpdateFirmwareResponse.json schema.

tropxy replied:
I am not sure that was labeled as final version, but the one I got, was the one officially released
[OCPP 1.6 security whitepaper edition 2.pdf](https://github.com/mobilityhouse/ocpp/files/6435221/OCPP.1.6.security.whitepaper.edition.2.pdf)

villekr commented:
I will also sort payload classes in call.py and call_result.py based on communication direction as are previous payloads defined. Also enums.py is using alphabetical sorting so I'll follow that as well.

villekr commented:
I made suggested/requested changed (hopefully all this time) and also rebased against master.

tropxy reviewed (approved):
Looks good now, to me ;)

villekr commented:
@OrangeTux @tropxy When are you planning to create next release that will include this PR?

villekr commented:
@OrangeTux @tropxy Is there something still regarding this PR that you would like to see changed? When are you planning to include this in release?

tropxy commented:
As I said last time, for me this PR seems complete. There are some lack of new lines in the schema files, but that is not a blocker for me at least. @OrangeTux needs to take an action here and check if he confirms that the PR is ready to merge. Without his feedback, your PR cant be merged
